### ISSUE TITLE
Improve startup speed for CLI applications

### ISSUE THREAD

**Issue by bobtiernay-okta** [micronaut-projects/micronaut-core#889]
### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

```
mn create-cli-app test --build=maven && cd test && ./mvnw package && time java -jar target/test-0.1.jar
```

### Expected Behaviour

Expect startup time to be under 500ms.

### Actual Behaviour

```
time java -jar target/test-0.1.jar
17:21:14.173 [main] INFO  io.micronaut.runtime.Micronaut - No embedded container found. Running as CLI application
java -jar target/test-0.1.jar  2.22s user 0.22s system 177% cpu 1.374 total
```

#### Profiler
![image](https://user-images.githubusercontent.com/25536705/48319321-82822a80-e5da-11e8-9e75-787770979edf.png)

#### Hotspots
![image](https://user-images.githubusercontent.com/25536705/48319327-975ebe00-e5da-11e8-8832-2ca20429f7cc.png)
![image](https://user-images.githubusercontent.com/25536705/48319355-fd4b4580-e5da-11e8-871b-dcdde327fe75.png)
![image](https://user-images.githubusercontent.com/25536705/48319368-1f44c800-e5db-11e8-8736-479f64ac4af8.png)

### Environment Information

- **Operating System**: macOS Darwin Kernel Version 17.4.0
- **Micronaut Version:** 1.0.0.RC2
- **JDK Version:** 1.8.0_181

### Example Application

https://github.com/bobtiernay-okta/test

**Comment by graemerocher** [comment]
@bobtiernay-okta I upgraded the app to 1.0

```diff
diff --git a/dependency-reduced-pom.xml b/dependency-reduced-pom.xml
index 866d101..83b7222 100644
--- a/dependency-reduced-pom.xml
+++ b/dependency-reduced-pom.xml
@@ -19,12 +19,12 @@
             <annotationProcessorPaths>
               <path>
                 <groupId>io.micronaut</groupId>
-                <artifactId>inject-java</artifactId>
+                <artifactId>micronaut-inject-java</artifactId>^M
                 <version>${micronaut.version}</version>
               </path>
               <path>
                 <groupId>io.micronaut</groupId>
-                <artifactId>validation</artifactId>
+                <artifactId>micronaut-validation</artifactId>^M
                 <version>${micronaut.version}</version>
               </path>
             </annotationProcessorPaths>
@@ -79,8 +79,8 @@
   <dependencies>
     <dependency>
       <groupId>io.micronaut</groupId>
-      <artifactId>inject-java</artifactId>
-      <version>1.0.0.RC2</version>
+      <artifactId>micronaut-inject-java</artifactId>^M
+      <version>1.0.0</version>^M
       <scope>provided</scope>
     </dependency>
     <dependency>
@@ -106,7 +106,7 @@
     <dependencies>
       <dependency>
         <groupId>io.micronaut</groupId>
-        <artifactId>bom</artifactId>
+        <artifactId>micronaut-bom</artifactId>^M
         <version>${micronaut.version}</version>
         <type>pom</type>
         <scope>import</scope>
@@ -117,7 +117,7 @@
     <jdk.version>1.8</jdk.version>
     <exec.mainClass>test.Application</exec.mainClass>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <micronaut.version>1.0.0.RC2</micronaut.version>
+    <micronaut.version>1.0.0</micronaut.version>^M
   </properties>
 </project>
 
diff --git a/pom.xml b/pom.xml
index edb0817..fc75ee3 100644
--- a/pom.xml
+++ b/pom.xml
@@ -5,7 +5,7 @@
   <version>0.1</version>
   <properties>
     <exec.mainClass>test.Application</exec.mainClass>
-    <micronaut.version>1.0.0.RC2</micronaut.version>
+    <micronaut.version>1.0.0</micronaut.version>
     <jdk.version>1.8</jdk.version>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
   </properties>
@@ -19,7 +19,7 @@
     <dependencies>
       <dependency>
         <groupId>io.micronaut</groupId>
-        <artifactId>bom</artifactId>
+        <artifactId>micronaut-bom</artifactId>
         <version>${micronaut.version}</version>
         <type>pom</type>
         <scope>import</scope>
@@ -29,7 +29,7 @@
   <dependencies>
         <dependency>
           <groupId>io.micronaut</groupId>
-          <artifactId>runtime</artifactId>
+          <artifactId>micronaut-runtime</artifactId>
           <scope>compile</scope>
         </dependency>
         <dependency>
@@ -39,22 +39,22 @@
         </dependency>
         <dependency>
           <groupId>io.micronaut.configuration</groupId>
-          <artifactId>picocli</artifactId>
+          <artifactId>micronaut-picocli</artifactId>
           <scope>compile</scope>
         </dependency>
         <dependency>
           <groupId>io.micronaut</groupId>
-          <artifactId>inject</artifactId>
+          <artifactId>micronaut-inject</artifactId>
           <scope>compile</scope>
         </dependency>
         <dependency>
           <groupId>io.micronaut</groupId>
-          <artifactId>validation</artifactId>
+          <artifactId>micronaut-validation</artifactId>
           <scope>compile</scope>
         </dependency>
         <dependency>
           <groupId>io.micronaut</groupId>
-          <artifactId>inject-java</artifactId>
+          <artifactId>micronaut-inject-java</artifactId>
           <scope>provided</scope>
         </dependency>
         <dependency>
@@ -130,12 +130,12 @@
             <annotationProcessorPaths>
               <path>
                 <groupId>io.micronaut</groupId>
-                <artifactId>inject-java</artifactId>
+                <artifactId>micronaut-inject-java</artifactId>
                 <version>${micronaut.version}</version>
               </path>
               <path>
                 <groupId>io.micronaut</groupId>
-                <artifactId>validation</artifactId>
+                <artifactId>micronaut-validation</artifactId>
                 <version>${micronaut.version}</version>
               </path>
 </annotationProcessorPaths>
```

Then for CLI apps for the fastest startup time we recommend `-XX:TieredStopAtLevel=1 -noverify`. This results in:

```
time java -XX:TieredStopAtLevel=1 -noverify -jar target/test-0.1.jar 
10:08:15.610 [main] INFO  io.micronaut.runtime.Micronaut - No embedded container found. Running as CLI application

real	0m0.615s
user	0m0.740s
sys	0m0.128s
```

Which is pretty fast for a Java app. If you can compute it into Graal obviously it could be made even better. There are probably things we can improve further. We will investigate

**Comment by bobtiernay-okta** [comment]
Hey @graemerocher. Thanks for taking a look.

For my production scripts, I have the same JVM optimizations as yours. However, when I compare the runtime against Guice + Picocli, it's still about twice as slow. Considering that Guice is reflection based, I would have assumed that Micronaut would have won the day. The majority of time seems to be spent setting up the application  context. I'm wondering if there are things like disabling Caffeine or other unused code paths that could be helpful. Any thoughts here?

Btw, I upgraded the project as well with about similar results as previous.

Graal is definitely something I have my eye on, but for technical reasons I cannot use ATM.

**Comment by graemerocher** [comment]
@bobtiernay-okta yes there are some internal caches and also Jackson (which takes 200ms) that gets created it seems. We also seem to be loading the distributed configuration bootstrap context for the CLI which seems for CLI apps. We could definitely optimise this more, will take a look. Thanks for the feedback.

**Comment by graemerocher** [comment]
@bobtiernay-okta bit of a hack but this shaves off a bit of time too by disabling the bootstrap context (happens for server less functions):

```
time java -Dmicronaut.environments=function -XX:TieredStopAtLevel=1 -noverify -jar target/test-0.1.jar 
14:19:20.753 [main] INFO  i.m.context.env.DefaultEnvironment - Established active environments: [function]
14:19:20.957 [main] INFO  io.micronaut.runtime.Micronaut - No embedded container found. Running as CLI application

real	0m0.512s
user	0m0.616s
sys	0m0.103s
```

**Comment by graemerocher** [comment]
https://github.com/micronaut-projects/micronaut-core/commit/7db8f8108d4db25d9a680bad929c37a9a60e37b4 removes a couple of hundred milliseconds

**Comment by bobtiernay-okta** [comment]
Anyway to make the use of `Caffine` in  `DefaultClassPathResourceLoader` optional / swap out the implementation? 

```java
    public DefaultClassPathResourceLoader(ClassLoader classLoader, String basePath) {
        this.isDirectoryCache = Caffeine.newBuilder().maximumSize(50L).build();
        this.classLoader = classLoader;
        this.basePath = this.normalize(basePath);
    }
```

It seems to be adding about 8-9% to the startup time using 1.0.2:

![image](https://user-images.githubusercontent.com/25536705/50567798-30769f00-0d17-11e9-8a9f-4cdde7a9afb6.png)

**Comment by graemerocher** [comment]
We will consider it, there aren't many concurrent linked hash map implementations around.

**Comment by bobtiernay-okta** [comment]
The issue isn't the map implementation, but rather the cost of the `NodeFactory` initialization. It just seems very heavy.

**Comment by graemerocher** [comment]
Not sure how that initialisation can be avoided without switching from Caffeine to something else.

@ben-manes any thoughts on this issue? Anyway Caffeine init can be made any faster? Would switching back to concurrenthashmap be faster?

**Comment by ben-manes** [comment]
NodeFactory should just be a reflection lookup. It used to use method handles, but that was removed due to a IVM [memory leak](https://github.com/ben-manes/caffeine/issues/222). I don’t understand why it would be slow offhand, but would be willing to iterate on any feedback to speed it up.

There is my older [ConcurrentLinkedHashMap](https://github.com/ben-manes/concurrentlinkedhashmap) library (Java 6) as a possible fallback.

**Comment by ben-manes** [comment]
Can you please [upgrade](https://github.com/micronaut-projects/micronaut-core/blob/3f1ef9ce4cfbcc20e6a08a08eca94fdf142d713d/gradle.properties#L43) from v2.5.2 (May 2017) to the latest, 2.6.2 (Feb. 2018)? In [2.6.1](https://github.com/ben-manes/caffeine/releases/tag/v2.6.1) we started using dynamic class loading to shrink the jar size and you will probably need to update your graal reflection config.

Previously the logic was more naive,
```java
String name = // determine optimal entry instance from cache builder
switch (name) {
  case [name]: return new [name]();
  ...
}
```

That meant a lot of constant strings, and maybe graal is having a harder time loading them. In the current code it uses reflection, via ClassLoader#loadClass().

We do this all because there are many different configurations, each having different memory requirements. Instead of taking the worst case overhead, we use generated classes to only include the fields necessary. Since most applications only use a handful of configurations, this is typically a good tradeoff as jar bloat (unloaded classes) isn't often a concern.

**Comment by bobtiernay-okta** [comment]
Just a point of clarification, I'm not using Graal as part of the build. I'm using JIT.

**Comment by graemerocher** [comment]
Switching to CLHM for some of our internal maps indeed does shave between 100-200ms off startup time. I have pushed these changes to master https://github.com/micronaut-projects/micronaut-core/commit/e8421f108f0e2c12756d070d589016a590bfa064

@bobtiernay-okta let me know if you see an improvement

**Comment by bobtiernay-okta** [comment]
Awesome once again, thanks @graemerocher! 

I will certainly give this a spin when it hits a release to confirm.

**Comment by ben-manes** [comment]
Please be aware that CLHM inherits the non-atomic default `computeIfAbsent` implementation (6 usages). The [fork in Groovy](
https://github.com/apache/groovy/blob/master/src/main/java/org/apache/groovy/util/concurrentlinkedhashmap/ConcurrentLinkedHashMap.java#L777) added it though I am not familiar with the code changes and aware by accident. You might want to sync with their copy.

If you still have a dependency on Caffeine, it would be great to update it if possible too.

**Comment by graemerocher** [comment]
We are using a version forked from Groovy

**Comment by ben-manes** [comment]
Can you please double check? I don’t see their modifications in your copy. I can’t vouch for them, though.

**Comment by graemerocher** [comment]
Thanks I will add that

**Comment by graemerocher** [comment]
@ben-manes I looked at add the Guava tests you mention above. Most pass but it seems the Groovy implementation does not pass the entrySet assertions and I get

```
junit.framework.AssertionFailedError: addAll(nonePresent) should throw

	at junit.framework.Assert.fail(Assert.java:57)
	a
```

Do you know if these are valid failures or it is to do with how I am configuring the suite?

**Comment by ben-manes** [comment]
Does my original one pass or is it a bug in Groovy’s fork?

I don’t recall the test suite having many options for edge cases like this and is a bit hard to debug at times. The configuration is what I use in Caffeine and would have expected to pass.

**Comment by graemerocher** [comment]
I have not tested the original one, will give it a go, if it does I imagine it is a bug in the Groovy fork. It seems related to calling `addAll(..)` on a returned `entrySet()` method. It seems the methods should throw `UnsupportedOperationException`? Unless I am misunderstanding the tests

**Comment by ben-manes** [comment]
Sorry, got to my computer. This is a bug in the original code, as I tend to implement optional methods. However `Map.entrySet()` disallows add, which is surprising since `ConcurrentHashMap` does support it (among others). The Guava tests assert no add unless you add `CollectionFeature.SUPPORTS_ADD` which will fail because `keySet` and `valueSet` clearly cannot. So it's a quirk in the Collections Framework and Guava's testlib.

```
/**
  * Returns a {@link Set} view of the mappings contained in this map.
  * The set is backed by the map, so changes to the map are
  * reflected in the set, and vice-versa.  If the map is modified
  * while an iteration over the set is in progress (except through
  * the iterator's own {@code remove} operation, or through the
  * {@code setValue} operation on a map entry returned by the
  * iterator) the results of the iteration are undefined.  The set
  * supports element removal, which removes the corresponding
  * mapping from the map, via the {@code Iterator.remove},
  * {@code Set.remove}, {@code removeAll}, {@code retainAll} and
  * {@code clear} operations.  It does not support the
  * {@code add} or {@code addAll} operations.
  *
  * @return a set view of the mappings contained in this map
  */
```

**Comment by ben-manes** [comment]
If you prefer and don't care about possible atomic computes, I can investigate a new release (need to get JDK6 working again!). Otherwise you can change the fork by removing the `EntrySet#add` method to comply with the spec.

I know CHM tended to deviate due to compatible with `Hashtable` which also supports `add`, and I was using CHM as a reference. I probably removed it from Caffeine due to adding Guava's testlib and realizing, and then forgetting about it afterwards.

**Comment by graemerocher** [comment]
@ben-manes Thanks for offering, the implementation currently does what we need we don't use the methods that are failing anywhere so I think it is ok to just leave it as is. We are unlikely to make any changes to the fork from here on.

**Comment by ben-manes** [comment]
Should we open an issue for upgrading Caffeine? I released 2.7 that includes an adaptive policy, which improves hit rates in recency-skewed workloads. I don't know the characteristics of yours, but it may be helpful.
